x/dclupgrade/handler_test.go

```diff
@@ -0,0 +1,432 @@
+// Copyright 2020 DSR Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//nolint:testpackage,lll, staticcheck
+package dclupgrade
+
+import (
+	"context"
+	"testing"
+
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+	testconstants "github.com/zigbee-alliance/distributed-compliance-ledger/integration_tests/constants"
+	testkeeper "github.com/zigbee-alliance/distributed-compliance-ledger/testutil/keeper"
+	"github.com/zigbee-alliance/distributed-compliance-ledger/testutil/testdata"
+	dclauthtypes "github.com/zigbee-alliance/distributed-compliance-ledger/x/dclauth/types"
+	"github.com/zigbee-alliance/distributed-compliance-ledger/x/dclupgrade/keeper"
+	"github.com/zigbee-alliance/distributed-compliance-ledger/x/dclupgrade/types"
+)
+
+type TestSetup struct {
+	T *testing.T
+	// Cdc         *amino.Codec
+	Ctx           sdk.Context
+	Wctx          context.Context
+	Keeper        *keeper.Keeper
+	DclauthKeeper *DclauthKeeperMock
+	UpgradeKeeper *UpgradeKeeperMock
+	Handler       sdk.Handler
+	// Querier     sdk.Querier
+}
+
+func (setup *TestSetup) AddAccount(
+	accAddress sdk.AccAddress,
+	roles []dclauthtypes.AccountRole,
+) {
+	for _, role := range roles {
+		setup.DclauthKeeper.On("HasRole", mock.Anything, accAddress, role).Return(true)
+	}
+	setup.DclauthKeeper.On("HasRole", mock.Anything, accAddress, mock.Anything).Return(false)
+}
+
+func Setup(t *testing.T) TestSetup {
+	dclauthKeeper := &DclauthKeeperMock{}
+	upgradeKeeper := &UpgradeKeeperMock{}
+	keeper, ctx := testkeeper.DclupgradeKeeper(t, dclauthKeeper, upgradeKeeper)
+
+	setup := TestSetup{
+		T:             t,
+		Ctx:           ctx,
+		Wctx:          sdk.WrapSDKContext(ctx),
+		Keeper:        keeper,
+		DclauthKeeper: dclauthKeeper,
+		UpgradeKeeper: upgradeKeeper,
+		Handler:       NewHandler(*keeper),
+	}
+
+	return setup
+}
+
+func TestHandler_ProposedUpgradeExists(t *testing.T) {
+	setup := Setup(t)
+
+	trusteeAccAddress1 := testdata.GenerateAccAddress()
```

Comment:
Why do we need this setup code in every test (lines 78-85)? Can we move it to `Setup`?

Reply:
We define `DclauthKeeper.CountAccountsWithRole` return value right in a test. So for clarity it is better to have all the accounts being created right in a test too.